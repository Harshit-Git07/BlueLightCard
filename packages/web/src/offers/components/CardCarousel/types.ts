export interface CardCarouselOffer {
  offername?: string;
  companyname?: string;
  imageUrl: string;
  href: string;
  offerId?: number | string;
  companyId?: number | string;
  hasLink?: boolean;
  onClick?: () => void;
}

export interface CardCarouselProps {
  title?: string;
  itemsToShow: number;
  useSmallCards?: boolean;
  offers: CardCarouselOffer[];
}
