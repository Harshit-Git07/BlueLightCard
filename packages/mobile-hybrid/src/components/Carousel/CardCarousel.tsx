import { FC, SetStateAction, useState } from 'react';
import { Carousel } from 'react-responsive-carousel';
import 'react-responsive-carousel/lib/styles/carousel.min.css';
import { CardCarouselProps } from '@/components/Carousel/types';
import Card from '../Card/Card';
import { cssUtil } from '@/utils/cssUtil';

const CardCarousel: FC<CardCarouselProps> = ({ slides, onSlideItemClick, onSlideChanged }) => {
  const [currentSlide, setCurrentSlide] = useState(0);

  const handleSlideChange = (index: SetStateAction<number>) => {
    setCurrentSlide(index);
    if (onSlideChanged) {
      onSlideChanged(index as number);
    }
  };

  return (
    <Carousel
      showStatus={false} // Hide the status indicator
      showThumbs={false} // Enable thumbs, defaults to true
      showIndicators={false} // Show the slide dots
      showArrows={false} // Show the navigation arrows
      centerMode={true} // Center the current item
      selectedItem={currentSlide} // Set the active slide index
      preventMovementUntilSwipeScrollTolerance={true}
      swipeScrollTolerance={10}
      onChange={handleSlideChange} // Event handler for slide change
    >
      {slides.map((slide, index) => {
        const css = cssUtil([
          'transform transition-transform px-4',
          currentSlide !== index ? 'scale-[.9]' : '',
        ]);
        return (
          <div key={index} className={css}>
            <Card
              text={slide.title}
              title={slide.text}
              imageSrc={slide.imageSrc}
              onClick={() => onSlideItemClick && onSlideItemClick(slide.id)}
            />
          </div>
        );
      })}
    </Carousel>
  );
};

export default CardCarousel;
