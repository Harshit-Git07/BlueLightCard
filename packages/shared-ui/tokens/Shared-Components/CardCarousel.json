{
  "card-pagination-colour": {
    "light": {
      "value": "{colour.primary.light}",
      "type": "color"
    },
    "dark": {
      "value": "{colour.primary.dark}",
      "type": "color"
    }
  }
}
