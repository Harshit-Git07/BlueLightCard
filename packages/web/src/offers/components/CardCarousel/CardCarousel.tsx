import Carousel from '@/components/Carousel/Carousel';
import useIsVisible from '@/hooks/useIsVisible';
import Heading from '@/components/Heading/Heading';
import { useRef, useState, useEffect } from 'react';
import OfferCard from '@/offers/components/OfferCard/OfferCard';
import LoadingPlaceholder from '@/offers/components/LoadingSpinner/LoadingSpinner';
import { CardCarouselProps } from './types';

const CardCarousel = ({ title, itemsToShow, offers, useSmallCards }: CardCarouselProps) => {
  const carouselRef = useRef<HTMLDivElement>(null);
  const isVisible = useIsVisible(carouselRef);

  /** Set temporary placeholder height to only load carousel if space on screen */
  const [className, setClassName] = useState('h-[400px]');

  /** If element is visible on screen, remove placeholder height */
  useEffect(() => {
    if (isVisible) setClassName('');
  }, [isVisible]);

  return (
    <div className={className} ref={carouselRef}>
      {isVisible && (
        <>
          <Heading headingLevel="h2">{title}</Heading>
          {offers.length === 0 && <LoadingPlaceholder />}
          {offers.length > 0 && (
            <Carousel
              showControls
              autoPlay
              elementsPerPageDesktop={itemsToShow}
              elementsPerPageLaptop={itemsToShow}
              elementsPerPageMobile={1}
              elementsPerPageTablet={2}
            >
              {offers.map((offer, index) => (
                <OfferCard
                  key={index}
                  alt={'Card'}
                  offerName={offer.offername ?? ''}
                  companyName={offer.companyname ?? ''}
                  imageSrc={offer.imageUrl}
                  offerLink={offer.href}
                  variant={useSmallCards ? 'small' : ''}
                />
              ))}
            </Carousel>
          )}
        </>
      )}
    </div>
  );
};

export default CardCarousel;
