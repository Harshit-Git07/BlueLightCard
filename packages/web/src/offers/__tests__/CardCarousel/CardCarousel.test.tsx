import CardCarousel from '@/offers/components/CardCarousel/CardCarousel';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';

const offerData = [
  {
    offername: 'test',
    companyname: 'test',
    imageUrl: 'test.com',
    href: 'test.com',
  },
  {
    offername: 'test',
    companyname: 'test',
    imageUrl: 'test.com',
    href: 'test.com',
  },
  {
    offername: 'test',
    companyname: 'test',
    imageUrl: 'test.com',
    href: 'test.com',
  },
  {
    offername: 'test',
    companyname: 'test',
    imageUrl: 'test.com',
    href: 'test.com',
  },
  {
    offername: 'test',
    companyname: 'test',
    imageUrl: 'test.com',
    href: 'test.com',
  },
];

describe('CardCarousel', () => {
  it('should render with 3 elements visbile', () => {
    const component = renderer.create(
      <CardCarousel title="test" itemsToShow={3} offers={offerData} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with 1 element visbile', () => {
    const component = renderer.create(
      <CardCarousel title="test" itemsToShow={1} offers={offerData} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with small cards visbile', () => {
    const component = renderer.create(
      <CardCarousel title="test" itemsToShow={3} useSmallCards offers={offerData} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with a loading spinner when offers is empty', () => {
    const component = renderer.create(<CardCarousel title="test" itemsToShow={3} offers={[]} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('should call onCarouselInteracted when interacted with', () => {
  const onCarouselInteractedMock = jest.fn();

  const { getByTestId } = render(
    <CardCarousel
      title="test"
      itemsToShow={3}
      offers={offerData}
      onCarouselInteracted={onCarouselInteractedMock}
    />
  );

  const carouselElement = getByTestId('carousel');

  // Simulate interaction
  fireEvent.scroll(carouselElement);
  expect(onCarouselInteractedMock).toHaveBeenCalledTimes(1);
});
it('should apply placeholder height when carousel is not visible', () => {
  const component = renderer.create(<CardCarousel title="test" itemsToShow={3} offers={[]} />);

  const carouselInstance = component.root.findByType('div');
  expect(carouselInstance.props.className).toContain('h-[400px]');
});

jest.mock('@/hooks/useIsVisible', () => jest.fn());

it('should remove placeholder height when carousel becomes visible', () => {
  const useIsVisible = require('@/hooks/useIsVisible');
  useIsVisible.mockReturnValue(true);

  const { getByTestId } = render(<CardCarousel title="test" itemsToShow={3} offers={[]} />);

  const carouselElement = getByTestId('carousel');
  expect(carouselElement.className).not.toContain('h-[400px]');
});
