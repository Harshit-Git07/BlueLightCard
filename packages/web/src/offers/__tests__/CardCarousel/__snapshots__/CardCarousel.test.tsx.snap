// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCarousel should render with 1 element visbile 1`] = `
<div
  className="h-[400px]"
/>
`;

exports[`CardCarousel should render with 3 elements visbile 1`] = `
<div
  className="h-[400px]"
/>
`;

exports[`CardCarousel should render with a loading spinner when offers is empty 1`] = `
<div
  className="h-[400px]"
/>
`;

exports[`CardCarousel should render with small cards visbile 1`] = `
<div
  className="h-[400px]"
/>
`;
